modules/hivemq-edge-module-opcua/src/main/java/com/hivemq/edge/adapters/opcua/OpcUaClientConnection.java

```diff
@@ -0,0 +1,350 @@
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.stream.Collectors;
+
+import static com.hivemq.adapter.sdk.api.state.ProtocolAdapterState.ConnectionStatus.CONNECTED;
+import static com.hivemq.adapter.sdk.api.state.ProtocolAdapterState.ConnectionStatus.DISCONNECTED;
+import static com.hivemq.adapter.sdk.api.state.ProtocolAdapterState.ConnectionStatus.ERROR;
+import static com.hivemq.edge.adapters.opcua.Constants.PROTOCOL_ID_OPCUA;
+import static java.util.Objects.requireNonNull;
+import static org.eclipse.milo.opcua.stack.core.types.builtin.unsigned.Unsigned.uint;
+
+public class OpcUaClientConnection {
+    private static final @NotNull Logger log = LoggerFactory.getLogger(OpcUaClientConnection.class);
+
+    private final @NotNull String uri;
+    private final @NotNull List<OpcuaTag> tags;
+    private final @NotNull ProtocolAdapterTagStreamingService tagStreamingService;
+    private final @NotNull DataPointFactory dataPointFactory;
+    private final @NotNull EventService eventService;
+    private final @NotNull String adapterId;
+    private final @NotNull ProtocolAdapterState protocolAdapterState;
+    private final @NotNull ReentrantLock ioLock = new ReentrantLock();
+    private final @NotNull OpcUaSpecificAdapterConfig config;
+    private final @NotNull ProtocolAdapterMetricsService protocolAdapterMetricsService;
+
+    private volatile @Nullable OpcUaClientContext opcUaClientInstance;
+
+    public OpcUaClientConnection(
+            final @NotNull String uri,
+            final @NotNull List<OpcuaTag> tags,
+            final @NotNull OpcUaSpecificAdapterConfig config,
+            final @NotNull ProtocolAdapterTagStreamingService tagStreamingService,
+            final @NotNull DataPointFactory dataPointFactory,
+            final @NotNull EventService eventService,
+            final @NotNull ProtocolAdapterMetricsService protocolAdapterMetricsService,
+            final @NotNull String adapterId,
+            final @NotNull ProtocolAdapterState protocolAdapterState) {
+        this.uri = uri;
+        this.tags = tags;
+        this.config = config;
+        this.tagStreamingService = tagStreamingService;
+        this.dataPointFactory = dataPointFactory;
+        this.eventService = eventService;
+        this.adapterId = adapterId;
+        this.protocolAdapterState = protocolAdapterState;
+        this.protocolAdapterMetricsService = protocolAdapterMetricsService;
+    }
+
+    private static byte @NotNull [] convertPayload(
+            final @NotNull DataValue dataValue,
+            final @NotNull EncodingContext serializationContext) {
+        //null value, empty buffer
+        if (dataValue.getValue().getValue() == null) {
+            return Constants.EMPTY_BYTES;
+        }
+
+        final ByteBuffer byteBuffer = OpcUaToJsonConverter.convertPayload(serializationContext, dataValue);
+        final ByteBuffer rewind = byteBuffer.asReadOnlyBuffer().rewind();
```

Comment:
Why do we need another `ByteBuffer rewind`? We can directly get byte array from `byteBuffer`.